File: Basics/graph/graph-DFS-BFS.py
from collections import defaultdict

class Graph:
    def __init__(self):
        self.adjacency_list = defaultdict(list)

    def add_vertex(self, vertex):
        self.adjacency_list[vertex] = []

    def add_edge(self, vertex1, vertex2):
        self.adjacency_list[vertex1].append(vertex2)
        self.adjacency_list[vertex2].append(vertex1)

    def dfs(self, start_vertex, visited=None):
        if visited is None:
            visited = set()
        print(start_vertex)
        visited.add(start_vertex)

        for neighbor in self.adjacency_list[start_vertex]:
            if neighbor not in visited:
                self.dfs(neighbor, visited)

    def bfs(self, start_vertex):
        queue = [start_vertex]
        visited = set()

        while queue:
            current_vertex = queue.pop(0)
            print(current_vertex)
            visited.add(current_vertex)

            for neighbor in self.adjacency_list[current_vertex]:
                if neighbor not in visited and neighbor not in queue:
                    queue.append(neighbor)

# Example usage:
my_graph = Graph()
my_graph.add_vertex("A")
my_graph.add_vertex("B")
my_graph.add_vertex("C")
my_graph.add_edge("A", "B")
my_graph.add_edge("B", "C")

print("DFS:")
my_graph.dfs("A")

print("BFS:")
my_graph.bfs("A")

File: Basics/graph/graph-DFS-BFS.js
class Graph {
    constructor() {
      this.adjacencyList = new Map();
    }
  
    addVertex(vertex) {
      this.adjacencyList.set(vertex, []);
    }
  
    addEdge(vertex1, vertex2) {
      this.adjacencyList.get(vertex1).push(vertex2);
      this.adjacencyList.get(vertex2).push(vertex1);
    }
  
    dfs(startVertex, visited = new Set()) {
      console.log(startVertex);
      visited.add(startVertex);
  
      for (const neighbor of this.adjacencyList.get(startVertex)) {
        if (!visited.has(neighbor)) {
          this.dfs(neighbor, visited);
        }
      }
    }
  
    bfs(startVertex) {
      const queue = [startVertex];
      const visited = new Set();
  
      while (queue.length > 0) {
        const currentVertex = queue.shift();
        console.log(currentVertex);
        visited.add(currentVertex);
  
        for (const neighbor of this.adjacencyList.get(currentVertex)) {
          if (!visited.has(neighbor) && !queue.includes(neighbor)) {
            queue.push(neighbor);
          }
        }
      }
    }
  }
  
  // Example usage:
  const myGraph = new Graph();
  myGraph.addVertex("A");
  myGraph.addVertex("B");
  myGraph.addVertex("C");
  myGraph.addEdge("A", "B");
  myGraph.addEdge("B", "C");
  
  console.log("DFS:");
  myGraph.dfs("A");
  
  console.log("BFS:");
  myGraph.bfs("A");
  